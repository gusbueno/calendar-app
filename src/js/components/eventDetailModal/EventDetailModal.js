import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import styles from './EventDetailModal.scss'

class EventDetailModal extends PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      title: '',
      description: ''
    }

    this._handleTitleChange = this._handleTitleChange.bind(this)
    this._handleDescriptionChange = this._handleDescriptionChange.bind(this)
    this._saveEvent = this._saveEvent.bind(this)
    this._updateEvent = this._updateEvent.bind(this)
    this._deleteEvent = this._deleteEvent.bind(this)
  }

  componentDidMount () {
    this.props.isEditMode && this.setState({ title: this.props.data.title, description: this.props.data.description })
  }

  _handleTitleChange (e) {
    const title = e.target.value
    this.setState({ title })
  }

  _handleDescriptionChange (e) {
    const description = e.target.value
    this.setState({ description })
  }

  _saveEvent () {
    const { ts } = this.props.data
    const { title, description } = this.state
    const data = { ts, title, description }
    this.props.onSaveEvent(data)
  }

  _updateEvent () {
    const { id, ts } = this.props.data
    const { title, description } = this.state
    const data = { id, ts, title, description }
    this.props.onUpdateEvent(data)
  }

  _deleteEvent () {
    const { id } = this.props.data
    this.props.onDeleteEvent(id)
  }

  render () {
    const date = moment(this.props.data.ts).format('YYYY-MM-DD')
    return (
      <div className={styles['event-detail-modal']}>
        <div className={styles['panel']}>
          <div className={styles['panel-header']}>
            <span className={styles.date}>{date}</span>
            <span className={styles['close-btn']} onClick={this.props.onCloseEventDetail}>Close</span>
          </div>
          <div className={styles['panel-body']}>
            <div className={styles['group-input']}>
              <label>Title</label>
              <input type='text' value={this.state.title} onChange={this._handleTitleChange} />
            </div>
            <div className={styles['group-input']}>
              <label>Description</label>
              <textarea value={this.state.description} rows='10' onChange={this._handleDescriptionChange} />
            </div>
          </div>
          <div className={[styles['panel-footer'], this.props.isEditMode && styles['edit-mode']].join(' ')}>
            {this.props.isEditMode && <button type='button' className={styles.delete} onClick={this._deleteEvent}>Delete</button>}
            {this.props.isEditMode ? <button type='button' className={styles.save} onClick={this._updateEvent}>Update</button> : <button type='button' className={styles.save} onClick={this._saveEvent}>Save</button>}
          </div>
        </div>
      </div>
    )
  }
}

EventDetailModal.propTypes = {
  isEditMode: PropTypes.bool.isRequired,
  data: PropTypes.object.isRequired,
  onCloseEventDetail: PropTypes.func.isRequired,
  onSaveEvent: PropTypes.func.isRequired,
  onUpdateEvent: PropTypes.func.isRequired,
  onDeleteEvent: PropTypes.func.isRequired
}

export default EventDetailModal
