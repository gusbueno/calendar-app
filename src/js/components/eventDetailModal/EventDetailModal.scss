@import "../../../styles/variables";

.event-detail-modal {
  display: flex;
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background-color: rgba($color-black, 0.9);
  padding: 20px 50px;
  flex-direction: column;
  box-sizing: border-box;
  justify-content: center;
  align-items: center;

  .panel {
    width: 400px;
    height: 200px;

    &-header {
      padding: 10px;
      box-sizing: border-box;
      border-top-left-radius: 5px;
      border-top-right-radius: 5px;
      background-color: $color-bone;
      display: flex;
      justify-content: space-between;
      align-items: center;

      .date {
        font-size: 24px;
        font-family: 'Roboto-Bold';
      }

      .close-btn {
        font-size: 12px;
        font-family: 'Roboto-Regular';
        color: $color-green;
        &:hover {
          cursor: pointer;
        }
      }
    }

    &-body {
      padding: 10px;
      box-sizing: border-box;
      background-color: $color-alabaster;

      .group-input {
        display: flex;
        flex-direction: column;
        margin-bottom: 10px;

        label {
          font-size: 12px;
          font-family: 'Roboto-Regular';
          margin-bottom: 5px;
        }

        input, textarea {
          font-size: 14px;
          font-family: 'Roboto-Regular';
          border-radius: 5px;
          border: 1px solid $color-green;
          padding: 5px;
          box-sizing: border-box;
          &:focus {
            outline: none;
          }
        }
      }
    }

    &-footer {
      display: flex;
      padding: 10px;
      border-bottom-left-radius: 5px;
      border-bottom-right-radius: 5px;
      background-color: $color-white;
      justify-content: flex-end;
      align-items: center;
      &.edit-mode {
        justify-content: space-between;
      }

      .save {
        background-color: $color-green;
        color: white;
      }

      .delete {
        background-color: orange;
      }

      button {
        padding: 10px;
        box-sizing: border-box;
        font-size: 18px;
        font-family: 'Roboto-Regular';
        border: 0;
        border-radius: 5px;
        &:hover {
          cursor: pointer;
        }
        &:focus {
          outline: none;
        }
      }
    }
  }
}