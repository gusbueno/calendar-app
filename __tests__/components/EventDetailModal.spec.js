/* global describe, it, expect, jest */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import EventDetailModal from '../../src/js/components/eventDetailModal/EventDetailModal'

describe('<EventDetailModal /> Component', () => {
  it('should match with the snapshot', () => {
    const tree = renderer.create(
      <EventDetailModal
        isEditMode={false}
        data={{ ts: 601293600 }}
        onCloseEventDetail={() => {}}
        onSaveEvent={() => {}}
        onUpdateEvent={() => {}}
        onDeleteEvent={() => {}}
      />
    )
    const json = tree.toJSON()
    expect(json).toMatchSnapshot()
  })

  const onCloseEventDetailMock = jest.fn()
  const onSaveEventMock = jest.fn()
  const onUpdateEventMock = jest.fn()
  const onDeleteEventMock = jest.fn()
  const eventDetailModalShallow = shallow(
    <EventDetailModal
      isEditMode={false}
      data={{ ts: 601293600 }}
      onCloseEventDetail={onCloseEventDetailMock}
      onSaveEvent={onSaveEventMock}
      onUpdateEvent={onUpdateEventMock}
      onDeleteEvent={onDeleteEventMock}
    />
  )

  it('should exists', () => {
    expect(eventDetailModalShallow.length).toEqual(1)
  })

  it('should set title value into the state on title input change', () => {
    const e = { target: { value: 'meetup' } }
    eventDetailModalShallow.find('input').first().simulate('change', e)
    expect(eventDetailModalShallow.state().title).toEqual('meetup')
  })

  it('should set description value into the state on description input change', () => {
    const e = { target: { value: 'PWA vs Hybrid vs Native' } }
    eventDetailModalShallow.find('textarea').first().simulate('change', e)
    expect(eventDetailModalShallow.state().description).toEqual('PWA vs Hybrid vs Native')
  })

  it('should call onSaveEvent on click save button', () => {
    eventDetailModalShallow.find('.save').simulate('click')
    expect(onSaveEventMock).toHaveBeenCalled()
  })

  it('should call onCloseEventDetail on click close button', () => {
    eventDetailModalShallow.find('.close-btn').simulate('click')
    expect(onCloseEventDetailMock).toHaveBeenCalled()
  })

  const eventDetailEditModeShallow = shallow(
    <EventDetailModal
      isEditMode
      data={{ id: 1, title: 'meetup', description: 'PWA vs Hybrid vs Native', ts: 601293600 }}
      onCloseEventDetail={onCloseEventDetailMock}
      onSaveEvent={onSaveEventMock}
      onUpdateEvent={onUpdateEventMock}
      onDeleteEvent={onDeleteEventMock}
    />
  )

  it('should set title and description values to state on componentDidMount', () => {
    expect(eventDetailEditModeShallow.state().title).toEqual('meetup')
    expect(eventDetailEditModeShallow.state().description).toEqual('PWA vs Hybrid vs Native')
  })

  it('should call onUpdateEvent on click update button', () => {
    eventDetailEditModeShallow.find('.save').simulate('click')
    expect(onUpdateEventMock).toHaveBeenCalled()
  })

  it('should call onDeleteEvent on click delete button', () => {
    eventDetailEditModeShallow.find('.delete').simulate('click')
    expect(onDeleteEventMock).toHaveBeenCalled()
  })
})
