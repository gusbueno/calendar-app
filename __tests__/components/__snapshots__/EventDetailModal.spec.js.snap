// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventDetailModal /> Component should match with the snapshot 1`] = `
<div
  className="event-detail-modal"
>
  <div
    className="panel"
  >
    <div
      className="panel-header"
    >
      <span
        className="date"
      >
        1970-01-08
      </span>
      <span
        className="close-btn"
        onClick={[Function]}
      >
        Close
      </span>
    </div>
    <div
      className="panel-body"
    >
      <div
        className="group-input"
      >
        <label>
          Title
        </label>
        <input
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="group-input"
      >
        <label>
          Description
        </label>
        <textarea
          onChange={[Function]}
          rows="10"
          value=""
        />
      </div>
    </div>
    <div
      className="panel-footer false"
    >
      <button
        className="save"
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
